### PR TITLE
fix: tls key is not taken into account for target cluster

### PR DESCRIPTION
**Pull request by callaertanthony**
### Type of change

- Bugfix

### Description

I set this PR as bugfix because my understanding of the documentation is that the connexion from the proxy to a broker can handle an ssl key pair connexion.

Testing kroxylicious shows me that there was an SSL handshake failed error. After reading the source coude, I found that the client connexion doesn't handle the keys.

### Additional Context

To be able to connect to a kafka cluster through ssl.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [x] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
@callaertanthony welcome, thank you for the contribution, I'll be reviewing the PR today.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:163]: I think this is wrong as it assumes client will always present a client certificate.

I think the formulation you want will look like this:

```suggestion
                var sslContextBuilder = Optional.ofNullable(tls.key()).map(KeyProvider::forClient).orElse(SslContextBuilder.forClient());
```

**Comment by k-wall** [comment]
Could you add a integration test case to `TlsIT` that tests the case where Kroxylicious presents a client certificate to the Kafka Broker?

You won't be able to config the cluster that to use client auth using the `@Tls KafkaCluster cluster` annotation.  `KafkaClusterConfig#fromConstraints` doesn't support it.

You should be able to create the cluster programmatically instead and setup the key material by hand.

https://github.com/kroxylicious/kroxylicious-junit5-extension/blob/b635b5ef4e6fb2d350a2ad5399e51d974853181d/impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java#L385

**Comment by k-wall** [comment]
Could you also update the documentation?

https://github.com/callaertanthony/kroxylicious/blob/fix/client-tls-key/docs/deploying.adoc#upstream-tls

to illustrate the client auth use-case.

**Comment by k-wall** [comment]
Running `mvn process-test-sources` will correct this issue.  There's advice in the DEV_GUIDE about this and details about running the ITs locally.
 
Error:  Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.23.0:validate (validate-sources) on project kroxylicious-runtime: File '/home/runner/work/kroxylicious/kroxylicious/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java' has not been previously formatted. Please format file (for example by invoking `mvn -f kroxylicious-runtime net.revelc.code.formatter:formatter-maven-plugin:2.23.0:format`) and commit before running validation! -> [Help 1]

**Comment by callaertanthony** [comment]
Hi @k-wall,

What do you mean by ?
> tests the case where Kroxylicious presents a client certificate to the Kafka Broker?

Are you expecting a similar test as in https://github.com/callaertanthony/kroxylicious/blob/a9e321c38ebee28944d9f78a6d11a136f949af08/kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/config/tls/KeyPairTest.java#L25 ?

I didn't find any IT test providing client certificates, it seams that jks keystore is always generated for the IT tests, am I wrong ?
Which test strategy do you want ? Generate key pair chain from the server and provide it as client ? Is it ok to generate them and add them to resources ?

Regards

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java:53]: What are the performance implications of calling `new File(....)` here? How often do we expect this path to be invoked? 

The one nice benefit of this approach is it does allow the certificate to be rotated seamlessly. However I think we can get that in other ways to avoid the need to load the file each time. 

I'm happy to pick up caching the certificate files in a separate PR once this is merged.

**Review by SamBarker — COMMENTED** [review]
Thank you @callaertanthony for the PR.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java
Comment (callaertanthony) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java:53]: Hi @SamBarker,

I didn't test performance for the moment, as I know the OS already cache files, performance optimization should be between java and os.

I'm not yet fluent with kroxylicious implementation, but if I remember right my test, connection and auth is only validated at the first request, it may be better to have a concret strategy about hat, like the oauth2 workflow for example, with an expiration validation time and a revoked certificate (not validated by the server). Si as you mentioned, the cache in fact.

Happy to contribute

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java:53]: Once per connection sounds fine for now. We can revisit if profiling suggests the overheads are problematic.

**Comment by k-wall** [comment]
sorry for the delay getting back to you, I was out of the office.

> Hi @k-wall,
> 
> What do you mean by ?
> 
> > tests the case where Kroxylicious presents a client certificate to the Kafka Broker?
> 
> Are you expecting a similar test as in https://github.com/callaertanthony/kroxylicious/blob/a9e321c38ebee28944d9f78a6d11a136f949af08/kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/config/tls/KeyPairTest.java#L25 ?

that's the unit test.  thanks for adding that.  However, my comment was about adding an integration test where kroxylicious actually connects to a Kafka Cluster brought up under control of the test framework.

> 
> I didn't find any IT test providing client certificates, it seams that jks keystore is always generated for the IT tests, am I wrong ?

You are right.  There's no IT test case at the moment that currently creates a client keystore.  The ITs use the Junit5 extension/annotated parameters (I'm referring to `@Tls KafkaCluster cluster`) to inject the configured kafka clusters into the tests.  Unfortunately, the extension/annotation doesn't currently support the creation of a kafka cluster configured to use TLS client auth.

However, IT tests don't need to use the extension/annotated mechanism.  That's why I pointed you [kafkaClusterKraftModeSSL_ClientUsesSSLClientAuth](https://github.com/kroxylicious/kroxylicious-junit5-extension/blob/b635b5ef4e6fb2d350a2ad5399e51d974853181d/impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java#L385) of in the Extension's test.  It programmatically creates key material for the client and server and sets up the trust. It should be possible write a Kroxylicious IT that takes the same approach.

Please take a look.  If this doesn't make sense (we know KeytoolCertificateGenerator isn't too nice btw), I'd be happy to raise a PR against yours demonstrating what I mean.

> Which test strategy do you want ? Generate key pair chain from the server and provide it as client ? Is it ok to generate them and add them to resources ?

We've take the approach of programmatically generating the key material so far.  We'd prefer a consistency of approach.

> 
> Regards

Thanks again for the contribution.

File: docs/deploying.adoc
Comment (k-wall) [inline docs/deploying.adoc:240]: "client credentials" is ambiguous. After all, when using SSL client credentials flow over the wire too. How would you feel about?
```suggestion
The following illustrates connection to physical cluster using TLS client authentication (aka Mutual TLS).
```

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
@callaertanthony to move things along, I raised #890 to illustrate what I meant above.  Please take a look/leave review comments.  Thank you.

Can you re-base/add a CHANGELOG.md entry to this PR?  I'm then happy to merge this PR, and then merge #890 to follow it.

I'd love to see this all land on main soon.

File: docs/deploying.adoc
Comment (k-wall) [inline docs/deploying.adoc:255]: source block needs to be closed.
```suggestion
----
```

**Review by robobario — APPROVED** [review]
LGTM once Keith's docs suggestions are resolved

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
Going to merge this as it stands, my PR that will follow this will address the slight remaining coverage issue.  Thanks again @callaertanthony

**Pull request merged**
